runtime/major_gc.c

```diff
@@ -167,26 +292,58 @@ void caml_darken (value v, value *p /* not used */)
     CAMLassert (!Is_blue_hd (h));
     if (Is_white_hd (h)){
       ephe_list_pure = 0;
+      Hd_val (v) = Blackhd_hd (h);
       if (t < No_scan_tag){
-        Hd_val (v) = Grayhd_hd (h);
-        *gray_vals_cur++ = v;
-        if (gray_vals_cur >= gray_vals_end) realloc_gray_vals ();
-      }else{
-        Hd_val (v) = Blackhd_hd (h);
+        mark_entry me = {v, 0};
+        mark_stack_push(Caml_state->mark_stack, me, NULL);
       }
     }
   }
 }
 
+/* This function adds blocks in the passed heap chunk [heap_chunk] to the mark
+   stack. It may return 0 if doing so would cause the mark stack to grow more
+   than a quarter full. This is to lower the chance of triggering another
+   overflow, which would be wasteful. Subsequent calls will continue progress.
+   It will return 1 when the supplied chunk has no more range to redarken. */
+static int redarken_chunk(char* heap_chunk, struct mark_stack* stk) {
+  value* p = (value*)Chunk_redarken_start(heap_chunk);
+  value* end = (value*)Chunk_redarken_end(heap_chunk);
+
+  while (p <= end) {
+    header_t hd = Hd_op(p);
+
+    if( Is_black_hd(hd) && Tag_hd(hd) < No_scan_tag ) {
```

Comment:
In the PR description you wrote :

> Later when the marking loop exhausts the mark stack we iterate through chunks and redarken blocks with black headers that have any non-black fields.

Here it looks like you push all black objects.  Well, not quite, because of the optimization in `mark_stack_push`: if the object is small enough and contains no non-black fields, it will not be pushed in the end.  And checking for non-black fields in large objects may be bad for latency.  So, the implementation is OK, it's just the PR description that confused me.

Reply:
Ah yes, I can see how that's not clear in the description. The optimisation in `mark_stack_push` does cull a significant number of objects we would otherwise have on the mark stack. That and only replenishing a percentage of the mark stack aim to avoid overflowing again.